README.md
# 程序员手册 V1.1.4

## 简介

内置了十多个实用的离线中文手册，包括：Linux、PHP、Python、JS等,以及提供了搜索devdocs、dash/zeal上的文档的功能

手册具有以下三个特色：

- 内置多个手册均可离线使用，且90%以上是中文
- 可以自行添加手册
- 可以查询devdocs、dash/zeal上的文档

## 基本操作

- 列表界面：
  - 查看选中的内容 `⏎`/`🖱左键`
  - 发送选中的内容至上一个活动窗口 `⇪ + ⏎`/`🖱中键`
  - 上下选择 `⇧`/`⇩`
- 手册界面：
  - 切换列表和手册视图 `tab`/`🖱右键`
  - 发送选中的内容至上一个活动窗口 `⇪ + ⏎`/`🖱中键`
  - 搜索下一个关键词 `⏎`
  - 选中文本后翻译 `T`
  - 选中文本后收藏 `S`
- 退出插件 `esc`

## 关键字

 - `手册设置`/`ManualOptions` ： 调整程序员手册的功能
 - `dash`/`zeal` ：打开dash 或 zeal 进行搜索
 - `linux命令` `php函数` `python库` `C函数` `win32api` `vim命令` `git命令` `docker命令` `sql手册` `uToolsAPI` 等

## 预览

![tg6WQO.png](https://s1.ax1x.com/2020/06/07/tg6WQO.png)
![tgcqE9.png](https://s1.ax1x.com/2020/06/07/tgcqE9.png)
![tg6Gon.png](https://s1.ax1x.com/2020/06/07/tg6Gon.png)
![tg6tJ0.png](https://s1.ax1x.com/2020/06/07/tg6tJ0.png)
![tgaws1.png](https://s1.ax1x.com/2020/06/07/tgaws1.png)
![tg6dQU.png](https://s1.ax1x.com/2020/06/07/tg6dQU.png)
![tg6wyF.png](https://s1.ax1x.com/2020/06/07/tg6wyF.png)
![tg6DeJ.png](https://s1.ax1x.com/2020/06/07/tg6DeJ.png)

## 安装

请直接在插件中心进行安装即可，或将本项目打包成`upx`安装

[项目地址](https://github.com/fofolee/uTools-Manuals)

[插件发布页](https://yuanliao.info/d/356)

[外置手册教程](https://yuanliao.info/d/356/27)


## 手册一览

所有手册资源均来自网络，如有遗漏，请留言告知

- php手册   | [官方中文文档](https://www.php.net/download-docs.php)       
- linux手册 | [离线手册：linux-command@jaywcjlove](https://github.com/jaywcjlove/linux-command)  [原始数据：linuxde.net](http://man.linuxde.net/) 
  - tmux 快捷键 | [ryuzheng提供](https://github.com/ryuzheng/TmuxCheatSheet) [中文原版](https://gist.github.com/ryerh/14b7c24dfd623ef8edc7) [英文原版](https://gist.github.com/MohamedAlaa/2961058)
- python手册 | [官方文档一译翻译版](https://yiyibooks.cn/xx/python_352/library/index.html)
  - beautifulSoup | [官方中文文档](https://www.crummy.com/software/BeautifulSoup/bs4/doc/index.zh.html)
  - requests | [官方中文文档](https://2.python-requests.org//zh_CN/latest/user/quickstart.html)
  - scrapy | [scrapy入门教程中文翻译](https://github.com/marchtea/scrapy_doc_chs)
  - pillow | [pillow官方文档中文翻译](https://pillow-cn.readthedocs.io/zh_CN/latest/)
  - tkinter | [易学教程](https://www.e-learn.cn/content/python/1101684)
- PyQt5文档 | [PyQt5中文教程](https://github.com/maicss/PyQt5-Chinese-tutorial)
- C函数 | C语言函数速查（Knocker 2004.7.7 版本 0.5）
- win32api | 新编win32api大全.chm
- go | [@永远の明日](https://yuanliao.info/u/21504), [源文档](https://studygolang.com/)
- vim命令 | [awesome-cheatsheets](https://github.com/skywind3000/awesome-cheatsheets/blob/master/editors/vim.txt) 
- git命令 | [php中文网git开发手册](http://www.php.cn/manual/view/34942.html) 
- sql手册 | [php中文网sql参考手册](http://www.php.cn/manual/view/21301.html) 
- uTools API | [uTools官方文档](https://u.tools/docs/developer/api.html) 
- java | [fondme.cn](https://blog.fondme.cn/apidoc/jdk-1.8-google/) 
- javascript | [MDN Web Docs](https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Index) 
- jQuery | [jQuery中文网](https://www.jquery123.com/) 
- vue | [官方中文文档](https://cn.vuejs.org/v2/api/) 
- payload | 收集自网络
- devdocs | [devdocs](https://devdocs.io/)



src/CHANGELOG.md
### v1.1.4

- 兼容暗黑模式
- 修复某些情况下导航目录被遮挡的 bug
- 修复按 S 搜藏时无法跳转到备忘(收藏)插件的 bug
- linux 收手册下新增一个 tmux 快捷键 [ryuzheng提供](https://github.com/ryuzheng/TmuxCheatSheet) [中文原版](https://gist.github.com/ryerh/14b7c24dfd623ef8edc7) [英文原版](https://gist.github.com/MohamedAlaa/2961058)


### v1.1.3

- 新增 go 文档，来自网友[@永远の明日](https://yuanliao.info/u/21504), [源文档](https://studygolang.com/)

### v1.1.2

- 新增功能
  - 新增在文档界面显示导航目录，方便在篇幅较长文档中快速定位，支持目录搜索。默认情况下，如果`h1+h2+h3`的数量大于10则自动显示目录

![tgaws1.png](https://s1.ax1x.com/2020/06/07/tgaws1.png)

  - 新增`win32api`手册

- 体验改善
  - 为了减少使用语音面板进行划词搜索时的干扰项，更改了开启`主输入框搜索`时匹配的正则，现在的匹配规则是`/^[a-zA-Z._]{2,20}$/i`
  - 优化列表搜索速度

- bug 修复
  - 修复了从`主输入框搜索`进入插件时未对关键词进行搜索的bug

### v1.1.1

- 新增linux下调用zeal
- 修复一个通过主输入框搜索的bug
- 增加一个快捷键提示，通过shift+enter也可以发送文本

### v1.1.0

- 适配新版uTools
- 更新`uToolsAPI`里的文档至最新版

### v1.0.3

- 增加收藏的功能，调用官方的`收藏`插件进行收藏，手册界面选择文本后按`S`键即可收藏

- 调整划词翻译功能的UI

- 更新列表界面和手册界面子输入框内的说明，提示各个快捷键的功能

- 修复一个BUG，该BUG会导致第一次载入插件时无法调用`dash`或`zeal`的文档，以及在设置手册时搜索关键字后无法跳转

![YAQV9f.gif](https://s1.ax1x.com/2020/05/06/YAQV9f.gif)

### v1.0.2

- 增加划词翻译的功能，调用有道的英译中的接口。手册界面选中文本后按`T`键即可翻译

### v1.0.1

- 一些bug修复、、

### v1.0.0

- 上架 uTools 插件商店,算是正式版了,去除了内置的插件更新提示
- 修复两个滚动事件引起的BUG
- (~~似乎~~)修复了mac下退出后再次进入无法触发鼠标单击事件的BUG
- 修复`payload.json`文件会被某些杀毒软件误报的情况,现在已做加密处理

### v0.0.4

​	抱歉由于这段时间私事太多，插件疏于更新。感谢@Xinu在插件无法使用的时候帮忙发布了修复版本。

- 修复了uTools更新后进入插件空白的BUG
- 添加Linux支持
- 修复了退出手册后再次进入无法搜索的BUG，但是MacOS下任然存在一个诡异的BUG，退出后再次进入无法触发鼠标单击事件，建议勾选**隐藏即完全退出**选项，windows不受影响
- 修复了添加外置手册时使用多关键词会出现问题的BUG
- 现在外置的Java手册有图标了

### v0.0.3

又有一大波更新来袭

先做一个名词解释,本文中所提及的

**手册界面**为：

![UTOOLS1557242800409.png](https://i.loli.net/2019/05/07/5cd1a3b058870.png)

**列表界面**为：

![UTOOLS1557242777511.png](https://i.loli.net/2019/05/07/5cd1a399a2c46.png)


##### 功能更新

- 添加了快速将内容发送至活动窗口的功能，在列表界面按`shift+enter`或鼠标中键即可将选中的函数名称发送至上一个系统的活动窗口，在手册界面，选中相关文本内容后，按`shift+enter`或鼠标中键即可发送。查完api后再也不用慢悠悠地手动复制函数了~~

- 集成了`devdocs`，在`手册设置`页面，点击下方的英文手册按钮即可进入`devdocs`的配置页面

![UTOOLS1557237205596.png](https://i.loli.net/2019/05/07/5cd18dd58ced8.png)

​    需要先在该页面点击下载按钮下载相应的文档目录，之后便可和内置手册一样使用

![UTOOLS1557240920471.png](https://i.loli.net/2019/05/07/5cd19c58614cd.png)

![UTOOLS1557240455149.png](https://i.loli.net/2019/05/07/5cd19a8732d26.png)

![UTOOLS1557240886582.png](https://i.loli.net/2019/05/07/5cd19c3659474.png)

   需要注意的是`devdocs`所有文档的`api`列表中均无中文注释，且只有目录文件是离线存储的，访问具体的手册页面需要联网

- 现在`手册设置`页面可以在顶端输入框内搜索快速查找需要的文档

- 添加快速启动`dash`或者`zeal`的功能，通过关键字`dash`即可进入

![UTOOLS1557241588196.png](https://i.loli.net/2019/05/07/5cd19ef3eba37.png)

​    注意，不同于`devdocs`，最后会根据查询内容调用`dash`或者`zeal`进行查看，故需要先安装`dash`或者`zeal`才可以使用此功能

- 添加了多关键词查找的功能，现在想怎么搜就怎么搜~

![UTOOLS1557241948612.png](https://i.loli.net/2019/05/07/5cd1a05ca1521.png)

- 添加了一个内置手册`PyQt5`，同时在内置的`python`手册中添加了多个`tkinter`的实例

  添加了一个内置手册`payload`，收录了一些网络安全测试的常用`payload`，没有手册页面，主要配合新出的发送文本到窗口的功能，实现快速输入`payload`，目前收录的比较少，后续会陆续补充
  

![UTOOLS1557242094634.png](https://i.loli.net/2019/05/07/5cd1a0ee86d07.png)

- 增加外置手册时，如需自定义`css`，不再是添加单个`css`文件，而是统一放入`assets`目录中，相关外置手册说明同步更新

##### 用户体验改善

- 当在列表界面进行搜索后，点击进入手册界面，会清空顶端输入框内的值
- 在某些情况下（例如读取外置手册失败时）会输出一些报错信息

##### bug修复

- 修复了添加外置手册后滚动条没有调整的bug，以及某些时候滚动条未自动调整的bug
- 修复检测更新功能中，即使点了不再提醒，下次重启`uTools`时仍会再次出现的bug

##### 本项目现已在Github开源，地址见下载一栏，欢迎提交pr，大家一同完善本项目，如果喜欢也请不吝star~

### v0.0.2

##### 功能更新

- 增加外置手册的功能，需要有一定的编写爬虫脚本的能力，详情[戳我](https://yuanliao.info/d/356/27)
![Snipaste_2019-04-20_10-45-20.png](https://i.loli.net/2019/04/20/5cba8859b3a66.png)
![Snipaste_2019-04-20_10-46-11.png](https://i.loli.net/2019/04/20/5cba8863a5266.png)
- 增加了三个内置手册`javascript`,`jQuery`,`vue`,一个外置手册`Java`。由于`Java`体积太大（目录10M，手册100M+），会使整个插件体积暴增，故采取外置形式，同时也作为外置手册的一个案例作为参考。
- 增加了版本检测的功能，当有新版本时会弹窗提示
![Snipaste_2019-04-20_14-04-40.png](https://i.loli.net/2019/04/20/5cbab6a27136e.png)
- 在设置页面增加了两个功能键`全部启用`和`全部禁用`

##### 用户体验改善

- 优化搜索结果的排序，现在的排序规则是，置顶全字匹配的内容，优先显示名称匹配的内容，其次显示描述匹配的内容
  ![Snipaste_2019-04-19_21-58-45.png](https://i.loli.net/2019/04/19/5cb9d4282b7a1.png)
- 滚动条样式调整，现在滚动条会自动隐藏，并在滚动时和鼠标移至滚动条位置时出现
- 设置页面的开关样式调整，现在开关的状态更加清晰明了
- linux、C等语言的手册现在支持语法高亮了
- 其他一些界面微调

##### BUG修复

- 修复了第一次进入插件时、列表更新时、方向键选择列表时，鼠标所在位置会被选中的BUG
- 修复了方向键进行列表选择时，没有到最后一行就向下滚动的BUG

### v0.0.1

相较于之前两版，做了比较大的更新：

- 整合所有手册到一个插件当中，现在有多达**九个**语言或工具的手册
- 添加了一个配置页面，可以选择需要启动的功能（注：默认情况下所有手册均未启用，需要先通过`手册设置`命令进行配置）
![Snipaste_2019-04-08_20-46-17.png](https://i.loli.net/2019/04/08/5cab4e92369df.png)


- 支持直接在`uTools`主输入框进行快速搜索查询（或复制文本后5s内呼出`uTools`），可在配置页面选择开启
![Snipaste_2019-04-08_20-47-13.png](https://i.loli.net/2019/04/08/5cab4f6fac77e.png)

- 子输入框功能增强。

  在**列表界面**用来搜索函数/命令，并高亮匹配文本
![Snipaste_2019-04-08_20-50-09.png](https://i.loli.net/2019/04/08/5cab4f9789d08.png)

  在**手册界面**则可以进行当前文档内容的搜索（回车键跳转到匹配文本位置）
![Snipaste_2019-04-08_21-10-30.png](https://i.loli.net/2019/04/08/5cab4fa84d2ac.png)

- 添加了一个快捷键`TAB`，以在列表界面和手册界面之间进行切换

- 现在支持点击手册里的外部链接跳转到相应网址（通过默认浏览器打开，之前并不支持，不知道你们发现没~~）

- 界面微调

- 一些BUG修复

src/assets/index.js
// html实体编码
escapeHtml = s => {
    return s.replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&apos;");
}

// 显示列表
showList = (text, index, listnum) => {
    window.manualVars.infoRows = [];
    var topRows = [],
        tailRows = [];
    if(text) text = text.toUpperCase()
    index.forEach(i => {
        let desc = i.desc != undefined ? i.desc : "",
            displayName = escapeHtml(i.name),
            displatDesc = escapeHtml(desc),
            displayType = i.type != undefined ? escapeHtml(i.type) : "",
            upName = i.name.toUpperCase(),
            upDesc = desc.toUpperCase(),
            initial = upName.slice(0, 1),
            topRow = true,
            matched = true;
        // 多关键词搜索
        text.trim().split(' ').forEach(keyword => {
            upName.includes(keyword) || (topRow = false);
            (upName.includes(keyword) || upDesc.includes(keyword)) || (matched = false);
        })
        var list = `<div class='info' path='${i.path}'>
        <div class="initial">${initial}</div>
        <div class="type">${displayType}</div>
        <div class="name">${displayName}</div>
        <div class="description">${displatDesc}</div>
        </div>`
        // 排序规则：置顶全字匹配，优先显示名称匹配
        topRow && ((upName == text) && topRows.unshift(list) || topRows.push(list)) || (matched && tailRows.push(list));
    });
    window.manualVars.infoRows = topRows.concat(tailRows);
    $("#mainlist").html(window.manualVars.infoRows.slice(0, listnum).join(''));
    let num = $(".info").length
    utools.setExpendHeight(num > 11 ? 550 : 50 * num);
    $(".select").removeClass('select');
    $(".info:first").addClass('select');
    // $('html').getNiceScroll().resize();
    // 鼠标锁，方式鼠标抢占选择条
    window.manualVars.mouseLockTime = new Date().getTime();
}

// 列表界面子输入框
mainlistSubInput = () => {
    utools.removeSubInput();
    utools.setSubInput(({ text }) => {
        // 列表搜索
            showList(text, index, 100);
            // 高亮结果
            text.split(' ').forEach(keyword => {
                keyword && $(".name,.description").highlight(keyword, 'listFounds');
            });
    }, '空格 多关键词搜索；中键发送选中条目到活动窗口');
}

// 手册界面子输入框
manualSubInput = () => {
    utools.removeSubInput();
    utools.setSubInput(({ text }) => {
        highlightManual("#manualBody", text);
    }, '搜索全文；选中文本：中键发送，T翻译，S收藏；Tab切换界面');
}


// 显示手册
showManual = path => {
    utools.setExpendHeight(550);
    // $('#manual').getNiceScroll().resize()
    if (/^((ht|f)tps?):\/\//.test(path)) {
        window.open(path);
    } else {
        let e = /(.*?)(\.html)*#(.*)/.exec(path);
        if (e) {
            var f = e[1] + '.html';
            var id = '#' + e[3];
        } else {
            var f = window.manualVars.dirs.idxFile ? path : path + '.html';
        }
        var file = `${window.manualVars.dirs.docPath}/${f}`;
        $("#mainlist").fadeOut();
        $(".load").html('Loading').show();
        var request = $.ajax({
            url: file,
            type: "GET",
        });
        request.done(data => {
            $(".load").hide();
            var relPath = f.substr(0, f.lastIndexOf('/') + 1);
            // a 标签改为相对路径
            data = data.replace(/(a.*?)href="(?!http)([^\#].*?)"/g, `$1href="${relPath}$2"`);
            // devdocs 语法高亮
            data = data.replace(/<pre.*?data-language="(.*?)">([\s\S]*?)<\/pre>/g, '<pre><code class="language-$1">$2</code></pre>')
            $("#manualBody").html(data)
            $("#manual").fadeIn()
            Prism.highlightAll();
            location.href = e ? id : '#manualBody';
            $('#manualNavi').autoMenu();
            if ($('h1,h2,h3').length < 10 && $('#manualNavi ul').is(":visible")) {
                $('.btn-box span').removeClass('icon-minus-sign').addClass('icon-plus-sign')
                // $('#manualBody').removeClass('withNaviBar')
                $('#manualNavi ul').hide()
            // } else {
                // $('#manualBody').addClass('withNaviBar')
                // $('#manualNavi input').focus()
            }
            manualSubInput();
        });
        request.fail(function (xhr, err) {
            $(".load").html('err').fadeOut();
            console.log(xhr, err);
        });
    }
}

// 手册/配置页搜索结果高亮
highlightManual = (selector, text) => {
    $(selector).removeHighlight('founds') ;
    if (text) {
        $(selector).highlight(text, 'founds');
        window.manualVars.findex = 0;
    }
}

// 初始化
// scrollInit = () => {
//     $("html").niceScroll({
//         enablekeyboard: false
//     });
//     $("#manual").niceScroll({
//         enablekeyboard: true
//     });
// }

// 向活动窗口发送文本
sendText = text => {
    window.copyTo(text);
    utools.hideMainWindow();
    window.paste();
}

// 切换列表和手册视图
toggleView = () => {
    if ($("#manual").is(":hidden") && $("#mainlist").is(":visible")) {
        $("#manual").fadeIn();
        $("#mainlist").fadeOut();
        utools.setExpendHeight(550);
        manualSubInput();
    } else if ($("#manual").is(":visible") && $("#mainlist").is(":hidden")) {
        $("#manual").fadeOut();
        $("#infopannel").fadeOut();
        $("#mainlist").fadeIn();
        let num = $(".info").length
        utools.setExpendHeight(num > 11 ? 550 : 50 * num);
        mainlistSubInput();
    }
}

// 继续加载内容
loadList = addnum => {
    if ($('#manual').is(':hidden') && $("#mainlist").is(":visible") && window.manualVars.infoRows) {
        var listnum = $(".info").length;
        if ($(window).scrollTop() >= (listnum * 50 - 550)) {
            $("#mainlist").append(window.manualVars.infoRows.slice(listnum, listnum + addnum).join(''));
            // $('html').getNiceScroll().resize();
        }
    }
}

adaptDarkMode = () => {
    if (utools.isDarkColors()) {
        !$('#darkmode').length && $('head').append('<link id="darkmode" rel="stylesheet" href="assets/styles/darkmode.css">')
            && $('#darkprism').attr('href', 'assets/styles/prism-dark.css')
    } else {
        $('#darkmode').length && $('#darkmode').remove() && $('#darkprism').attr('href', 'assets/styles/prism.css')
    }
}

// 进入插件
utools.onPluginEnter(async ({ code, type, payload }) => {
    if (fofoCommon.isRunningAtFirstTime()) fofoCommon.showChangeLog()
    // scrollInit();
    adaptDarkMode()
    // checkUpdate();
    window.manualVars = {}
    if (code == 'options') {
        window.manualVars.defaultPage = 0;
        showOptions();
        utools.setSubInput(({ text }) => {
            highlightManual(".keyword", text);
        }, '输入关键词快速查找文档');
    } else if (code == 'dash') {
        utools.setExpendHeight(0);
        utools.setSubInput(({ text }) => {
            window.manualVars.dashQuery = text;
        }, '输入关键词进行查询,例如 nodejs:fs');
    } else {
        $("#mainlist").fadeIn();
        var allFts = /^dd_/.test(code) ? await getDevdocs() : await getManuals(),
            baseDir,
            assetDir;
        switch (allFts[code].type) {
            case "default":
                baseDir = dirname;
                assetDir = `${baseDir}/assets/styles/${code}`
                window.manualVars.dirs = {
                    idxFile: `${baseDir}/index/${code}.json`,
                    docPath: `${baseDir}/docs`,
                }
                break;
            case "custom":
                baseDir = allFts[code].path;
                assetDir = `${baseDir}/assets`
                window.manualVars.dirs = {
                    idxFile: `${baseDir}/${code}.json`,
                    docPath: `${baseDir}`,
                }
                break;
            case "devdocs":
                assetDir = '';
                window.manualVars.dirs = {
                    docPath: allFts[code].url.slice(0, -11),
                }
                break;
        }
        // 自定义 CSS、JS情况下
        window.readDir(assetDir, (err, files) => {
            if (!err) {
                files.forEach(file => {
                    $('title').after(`<link rel="stylesheet" href="${assetDir}/${file}" name="manual">`)
                }) 
            } else {
                $('title').after('<link rel="stylesheet" href="assets/styles/manual.css" name="manual">')
            }
        })
        // 读取目录文件
        try {
            if (window.manualVars.dirs.idxFile) {
                index = await readFile(window.manualVars.dirs.idxFile);
                if (window.manualVars.dirs.idxFile.includes('payload.json')) {
                    index = JSON.parse(rc4(index, 'uTools'))
                } else {
                    index = JSON.parse(index);                   
                }
            } else {
                index = utools.db.get(code).data;
            }
            // 子输入框
            if ($('#manual').is(':hidden')) {
                mainlistSubInput();
            } else {
                manualSubInput();
            }
            if (type == 'regex') {
                utools.setSubInputValue(payload)
                // showList(payload, index, 300)
            } else {
                utools.setSubInputValue('')
                // showList('', index, 300)
            }
            if (type == 'window') {
                utools.hideMainWindow();
                copy();
                utools.showMainWindow();
                paste();
            }
        } catch(e) {
            $("#mainlist").html(e);
        }
    }
});

utools.onPluginOut(() => {
    $("#mainlist").empty().hide();
    $("#options").empty().hide();
    $("#manual").hide();
    $("#manualBody").empty()
    $("#manualNavi").empty()
    $(".load").empty().hide();
    $("#infopannel").empty().hide();;
    $('link[name="manual"]').remove();
})

// 单击列表，显示手册; 中键发送文本
$("#mainlist").on('mousedown', '.info', function (e) {
    if (1 == e.which) {
        var path = $(".select").attr('path');
        path && showManual(path);
    } else if (2 == e.which) {
        sendText($('.select .name').text());
    } else if (3 == e.which) {
        toggleView();
    }
});

// 鼠标滑过列表，高亮
$("#mainlist").on('mousemove', '.info', function () {
    // 鼠标锁 500ms
    var mouseUnlockTime = new Date().getTime();
    if (mouseUnlockTime - window.manualVars.mouseLockTime > 500) {
        $(".select").removeClass('select');
        $(this).addClass('select');
    }
});  

// 右键单击手册，退出手册; 中键发送文本
$("#manual").on('mousedown', function (e) {
    if (3 == e.which) {
        toggleView();
    } else if (2 == e.which) {
        var select = document.getSelection().toString();
        select && sendText(select);
    } else if (1 == e.which) {
        $("#infopannel").fadeOut(300);
    }
})

// 手册中a标签
$("#manual").on('mousedown', 'a', function (e) {
    if (1 == e.which) {
        var href = $(this).attr('href');
        /^\#/.test(href) || showManual(href);
    }
});

// 滚动到边界加载列表
$(document).scroll(() => {
    loadList(100);
})

// 按键监听
$(document).keydown(e => {
    switch (e.keyCode) {
        // TAB
        case 9:
            toggleView();
            break;
        // 回车
        case 13:
            // 按住shift键则发送文本
            if (event.shiftKey) {
                if ($('#manual').is(':hidden') && $("#mainlist").is(":visible")) {
                    sendText($('.select .name').text());
                } else if($('#manual').is(':visible') && $("#mainlist").is(":hidden")){
                    var select = document.getSelection().toString();
                    select && sendText(select);
                }
                return;
            } 
            // 列表界面进入手册
            if ($('#manual').is(':hidden') && $("#mainlist").is(":visible")) {
                var path = $(".select").attr('path');
                path && showManual(path);
            // 手册/配置界面搜索下一个
            } else if ($('.founds').length) {
                if ($('#manualNavi ul').is(":visible")) {
                    $('.btn-box span').removeClass('icon-minus-sign').addClass('icon-plus-sign')
                    $('#manualNavi ul').hide()
                }
                if (window.manualVars.findex > 0) {
                    $(`.founds:eq(${window.manualVars.findex - 1})`).removeClass('firstFound');
                } else {
                    $('.founds:last').removeClass('firstFound');  
                }
                $(`.founds:eq(${window.manualVars.findex})`).addClass('firstFound');
                $('.firstFound').get(0).scrollIntoView({ behavior: "smooth", block: "center" });
                if (window.manualVars.findex == $('.founds').length - 1) {
                    window.manualVars.findex = 0;
                } else {
                    window.manualVars.findex += 1;
                }
            // 快速启动 dash
            } else if (window.manualVars.dashQuery) {
                 window.dash(window.manualVars.dashQuery);
            }
            break;
        // 上
        case 38:
            if ($('#manual').is(':hidden') && $("#mainlist").is(":visible")) {
                let pre = $(".select").prev();
                // 没有到达边界时移动选择条
                if (pre.length != 0) {
                    if (pre.offset().top < $(window).scrollTop()) {
                        $("html").animate({ scrollTop: "-=50" }, 0);
                    }
                    pre.addClass("select");
                    $(".select:last").removeClass("select");
                    // 到达边界闪烁移动选择条
                } else {
                    $(".select").animate({ "opacity": "0.3" }).delay(500).animate({ "opacity": "1" })
                }
            }
            break;
        // 下
        case 40:
            if ($('#manual').is(':hidden') && $("#mainlist").is(":visible")) {
                let next = $(".select").next();
                // 没有到达边界时移动选择条
                if (next.length != 0) {
                    if (next.offset().top >= $(window).scrollTop() + 550) {
                        $("html").animate({ scrollTop: "+=50" }, 0);
                    }
                    loadList(100);
                    next.addClass("select");
                    $(".select:first").removeClass("select");
                    // 到达边界闪烁移动选择条
                } else {
                    $(".select").animate({ "opacity": "0.3" }).delay(500).animate({ "opacity": "1" })
                }
            }
            break;
        // 收藏
        case 83:
            if ($('#mainlist').is(':hidden') && $("#manual").is(":visible")) {
                let text = window.getSelection().toString();
                if (text) {
                    utools.redirect('备忘', { 'type': 'text', 'data': text })
                }
            }
            break;
        // 划词翻译
        case 84:
            if ($('#mainlist').is(':hidden') && $("#manual").is(":visible")) {
                let text = window.getSelection().toString();
                if (text) {
                    // if (/[\u4e00-\u9fa5]/.test(text)){
                        // utools.showNotification('中文你还看不懂嘛！', clickFeatureCode = null, silent = true)
                    // } else {
                        let enText = encodeURIComponent(text)
                        $("#infopannel").html('在线翻译中...').fadeIn(300);
                        $.get("http://fanyi.youdao.com/translate?&doctype=json&type=EN2ZH_CN&i=" + enText, data => {
                            let result = data.translateResult;
                            let cnText = '';
                            // 每段
                            for (var r of result) {
                                // 每句
                                for (var a of r) {
                                    cnText += a.tgt;
                                }
                                cnText += '<br>';
                            }
                            $("#infopannel").html(cnText)
                        })   
                    // }
                }
            }
            break;
    }
});